File: proxy-pattern/image-viewer/image-viewer.py
from abc import ABC, abstractmethod


# The interface that defines the common operations
class Image(ABC):
    @abstractmethod
    def display(self):
        pass


# The Real Subject which does the heavy lifting (e.g., loading a high-resolution image)
class RealImage(Image):
    def __init__(self, filename):
        self.filename = filename
        self._load_image_from_disk()

    def _load_image_from_disk(self):
        print(f"Loading image from disk: {self.filename}")

    def display(self):
        print(f"Displaying image: {self.filename}")


# The Proxy which acts as a placeholder
class ProxyImage(Image):
    def __init__(self, filename):
        self.filename = filename

        # RealImage will be created only when needed (to preserve resources)
        self.real_image = None

    def display(self):
        if self.real_image is None:
            # Lazy initialization
            self.real_image = RealImage(self.filename)
        self.real_image.display()


# Dryrun
if __name__ == "__main__":
    image = ProxyImage("shreyash.jpg")
    print("[DEBUG]: Image created. No loading yet. \n")

    image.display()  # Loads and displays the image

    print("\n[DEBUG]: Displaying again:")
    image.display()  # Doesn't load again, just displays the img

File: proxy-pattern/image-viewer/image-viewer.ts
interface Image {
  display(): void;
}

class RealImage implements Image {
  private filename: string;

  constructor(filename: string) {
    this.filename = filename;
    this.loadImageFromDisk();
  }

  display(): void {
    console.log(`Displaying image: ${this.filename}`);
  }

  loadImageFromDisk() {
    console.log(`Loading image from disk: ${this.filename}`);
  }
}

class ProxyImage implements Image {
  private filename: string;
  private realImage: RealImage | null;

  constructor(filename: string) {
    this.filename = filename;
    this.realImage = null;
  }

  display(): void {
    if (!this.realImage) {
      this.realImage = new RealImage(this.filename);
    }

    this.realImage?.display();
  }
}

// Dry run
const image = new ProxyImage("shreyash.png");
console.log("[DEBUG]: Image created, not loading yet.");
image.display();

console.log("\n[DEBUG]: Displaying again:");
image.display();
